File: browser-api/src/main/java/org/jetbrains/bio/browser/desktop/DesktopGenomeBrowser.java
package org.jetbrains.bio.browser.desktop;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import org.apache.log4j.Level;
import org.jdesktop.swingx.util.OS;
import org.jetbrains.bio.browser.AbstractGenomeBrowser;
import org.jetbrains.bio.browser.command.Command;
import org.jetbrains.bio.browser.command.Commands;
import org.jetbrains.bio.browser.headless.HeadlessGenomeBrowser;
import org.jetbrains.bio.browser.model.BrowserModel;
import org.jetbrains.bio.browser.model.LocationReference;
import org.jetbrains.bio.browser.model.ModelListener;
import org.jetbrains.bio.browser.model.SingleLocationBrowserModel;
import org.jetbrains.bio.browser.tracks.TrackView;
import org.jetbrains.bio.browser.util.LociCompletion;
import org.jetbrains.bio.genome.query.GenomeQuery;
import org.jetbrains.bio.util.Logs;

import javax.swing.*;
import java.awt.*;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author Roman.Chernyatchik
 *         <p>
 *         The basic usage looks like this:
 *         {@code new DesktopGenomeBrowser(...).show()}.
 *         One can set bookmarks previous to the {@link #show()} invocation.
 */
public class DesktopGenomeBrowser implements AbstractGenomeBrowser {
  public static final String GEMLBEE = "GeMLBee - Genome Multiple Locations Browser";
  private final List<TrackView> myTrackViews;
  private final Map<String, Function1<GenomeQuery, List<LocationReference>>> myLocationsMap;

  private MainPanel myPanel;
  private BrowserModel myBrowserModel;
  private ModelListener myModelListener;

  /**
   * This constructor specifies the available chromosomes and track views.
   *
   * @param query      Initial genome query
   * @param trackViews The list of the {@link TrackView} to display.
   */
  public DesktopGenomeBrowser(final GenomeQuery query, final List<TrackView> trackViews) {
    this(new SingleLocationBrowserModel(query), trackViews, LociCompletion.DEFAULT_COMPLETION);
  }

  public DesktopGenomeBrowser(final BrowserModel browserModel,
                              final List<TrackView> trackViews,
                              Map<String, Function1<GenomeQuery, List<LocationReference>>> locationsMap) {
    Logs.INSTANCE.addConsoleAppender(Level.INFO);
    Preconditions.checkArgument(!trackViews.isEmpty());
    myTrackViews = trackViews;
    myLocationsMap = locationsMap.keySet().stream().collect(Collectors.toMap(String::toLowerCase, locationsMap::get));
    myBrowserModel = browserModel;
    myModelListener = () -> getTrackViews().forEach(TrackView::fireRepaintRequired);
    myBrowserModel.addModelListener(myModelListener);
  }

  public void show() {
    preprocessTracks(myTrackViews, myBrowserModel.getGenomeQuery());

    myPanel = new MainPanel(this, ImmutableList.copyOf(myTrackViews));
    final JMenuBar menu = new JMenuBar();
    myPanel.fillMenu(menu);
    final JFrame frame = new JFrame(GEMLBEE);
    frame.setContentPane(myPanel);
    frame.setJMenuBar(menu);
    frame.invalidate();
    frame.validate();
    frame.repaint();
    frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
    frame.setLocation(0, 0);
    frame.setSize(!OS.isMacOSX() ? Toolkit.getDefaultToolkit().getScreenSize()
                                 : new Dimension(HeadlessGenomeBrowser.SCREENSHOT_WIDTH,
                                                 HeadlessGenomeBrowser.SCREENSHOT_HEIGHT));
    frame.setMinimumSize(new Dimension(500, 300));
    frame.setVisible(true);
    SwingUtilities.invokeLater(() -> myPanel.getTrackListComponent().requestFocus());
    BrowserSplash.close();
  }

  @Override
  public BrowserModel getBrowserModel() {
    return myBrowserModel;
  }

  @Override
  public void setBrowserModel(final BrowserModel model) {
    if (myBrowserModel.equals(model)) {
      return;
    }
    execute(Commands.createChangeModelCommand(this, model, (oldModel, newModel) -> {
      oldModel.removeModelListener(myModelListener);
      newModel.addModelListener(myModelListener);
      myBrowserModel = newModel;
      myPanel.setModel(newModel);
      myBrowserModel.modelChanged();
      return Unit.INSTANCE;
    }));
  }

  public List<TrackView> getTrackViews() {
    return myTrackViews;
  }

  @Override
  public void execute(final Command cmd) {
    // If command wasn't rejected:
    if (cmd != null) {
      getController().execute(cmd);
    }
  }

  public TrackListController getController() {
    return myPanel.getTrackListComponent().getTrackListController();
  }

  @Override
  public Map<String, Function1<GenomeQuery, List<LocationReference>>> getLocationsMap() {
    return myLocationsMap;
  }

  public void handlePositionChanged(final String text) {
    if (handleMultipleLocationsModel(text)) {
      return;
    }

    final SingleLocationBrowserModel model = (SingleLocationBrowserModel) getBrowserModel();
    final LocationReference locRef = LociCompletion.parse(text, model.getGenomeQuery());
    if (locRef != null) {
      try {
        execute(Commands.createGoToLocationCommand(model, locRef));
      } catch (final Exception e) {
        JOptionPane.showMessageDialog(myPanel.getParent(),
                                      e.getMessage(),
                                      "Go To Location",
                                      JOptionPane.ERROR_MESSAGE);
      }
    } else {
      JOptionPane.showMessageDialog(myPanel.getParent(),
                                    "Illegal location: " + text,
                                    "Go To Location",
                                    JOptionPane.ERROR_MESSAGE);
    }
  }
}

File: browser-api/src/main/kotlin/org/jetbrains/bio/browser/desktop/DesktopGenomeBrowser.kt
package org.jetbrains.bio.browser.desktop

import com.google.common.collect.ImmutableList
import org.apache.log4j.Level
import org.jdesktop.swingx.util.OS
import org.jetbrains.bio.browser.*
import org.jetbrains.bio.browser.headless.HeadlessGenomeBrowser
import org.jetbrains.bio.browser.model.BrowserModel
import org.jetbrains.bio.browser.model.LocationReference
import org.jetbrains.bio.browser.model.ModelListener
import org.jetbrains.bio.browser.model.SingleLocationBrowserModel
import org.jetbrains.bio.browser.query.desktop.DesktopInterpreter
import org.jetbrains.bio.browser.query.desktop.NewTrackViewListener
import org.jetbrains.bio.browser.query.desktop.TrackNameListener
import org.jetbrains.bio.browser.tracks.TrackView
import org.jetbrains.bio.genome.query.GenomeQuery
import org.jetbrains.bio.util.Logs
import java.awt.Dimension
import java.awt.Point
import java.awt.Toolkit
import java.util.*
import javax.swing.*

/**
 * The basic usage looks like this:
 *
 *     new DesktopGenomeBrowser(...).show()
 *
 * One can set bookmarks previous to the [.show] invocation.
 *
 * @author Roman Chernyatchik
 */
class DesktopGenomeBrowser(browserModel: BrowserModel,
                           override val trackViews: ArrayList<TrackView>,
                           locationsMap: Map<String, (GenomeQuery) -> List<LocationReference>>,
                           val interpreter: DesktopInterpreter)
:
        GenomeBrowser {

    override val locationsMap = locationsMap.mapKeys { it.key.toLowerCase() }

    private lateinit var mainPanel: MainPanel

    private var trackNameListeners: ArrayList<TrackNameListener> = ArrayList()

    fun addTrackNameListener(listener: TrackNameListener) {
        trackNameListeners.add(listener)
    }

    fun deleteTrackNameListener(listener: TrackNameListener) {
        trackNameListeners.remove(listener)
    }

    override var model: BrowserModel = browserModel
        set(value: BrowserModel) {
            if (value == model) {
                return
            }

            val oldModel = field
            field = value
            execute(oldModel.changeTo(field) { oldModel, newModel ->
                oldModel.removeListener(modelListener)
                newModel.addListener(modelListener)
                model = newModel
                mainPanel.setModel(newModel)
                model.modelChanged()
            })
        }

    val tracksCompletion: List<String>
        get() = interpreter.getAvailableNamedTracks()

    private val modelListener = object : ModelListener {
        override fun modelChanged() = trackViews.forEach(TrackView::fireRepaintRequired)
    }

    private val newTrackViewListener = object : NewTrackViewListener {
        override fun addNewTrackView(trackView: TrackView) {
            trackView.preprocess(browserModel.genomeQuery)
            trackViews.add(trackView)
            mainPanel.addTrackView(trackView)
            modelListener.modelChanged()
        }
    }

    /**
     * Just delegating...
     */
    private val newTrackNameListener = object : TrackNameListener {
        override fun addTrackName(name: String) {
            trackNameListeners.forEach { it.addTrackName(name) }
        }

        override fun deleteTrackName(name: String) {
            trackNameListeners.forEach { it.addTrackName(name) }
        }

    }

    init {
        Logs.addConsoleAppender(Level.INFO)
        require(trackViews.isNotEmpty())
        browserModel.addListener(modelListener)
        // listening for new track views...
        interpreter.addNewTrackViewListener(newTrackViewListener)
        interpreter.addNewTrackNameListener(newTrackNameListener)
    }

    override fun execute(command: Command?) {
        if (command != null) {
            controller.execute(command)
        }
    }

    val controller: TrackListController
        get() = mainPanel.trackListComponent.trackListController

    fun show() {
        if (OS.isMacOSX()) {
            // See http://stackoverflow.com/a/10366465/262432. This does not
            // affect the application name displayed to the left of the menu,
            // as explained here http://stackoverflow.com/a/26080591/262432.
            System.setProperty("apple.laf.useScreenMenuBar", "true")
        }

        BrowserSplash.display()

        preprocess()

        mainPanel = MainPanel(this, ImmutableList.copyOf(trackViews))
        val menu = JMenuBar()
        mainPanel.fillMenu(menu)
        JFrame(GEMLBEE).apply {
            contentPane = mainPanel
            jMenuBar = menu
            defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
            location = Point(0, 0)
            size = if (!OS.isMacOSX())
                Toolkit.getDefaultToolkit().screenSize
            else
                Dimension(HeadlessGenomeBrowser.SCREENSHOT_WIDTH,
                        HeadlessGenomeBrowser.SCREENSHOT_HEIGHT)
            minimumSize = Dimension(500, 300)
            isVisible = true
        }

        SwingUtilities.invokeLater { mainPanel.trackListComponent.requestFocus() }
        BrowserSplash.close()
    }

    fun handleAnyQuery(text: String) {
        try {
            handlePosition(text);
        } catch (e: Exception) {
            try {
                val msg = handleStatementQuery(text)
                if (msg.isNotEmpty()) {
                    JOptionPane.showMessageDialog(mainPanel.parent,
                            msg,
                            "Interpreter",
                            JOptionPane.INFORMATION_MESSAGE)
                }
            } catch (ne: Exception) {
                JOptionPane.showMessageDialog(mainPanel.parent,
                        "Can't parse position or statement: \n 1) ${e.message} \n 2) ${ne.message}",
                        "Go To Location/Interpreter",
                        JOptionPane.ERROR_MESSAGE)
            }
        }
    }

    fun handleOnlyPositionChanged(text: String) {
        if (interpreter.isParseable(text)) {
            return
        }
        try {
            handlePosition(text);
        } catch (e: Exception) {
            JOptionPane.showMessageDialog(mainPanel.parent,
                    "Can't parse position: \n ${e.message} \n",
                    "Go To Location",
                    JOptionPane.ERROR_MESSAGE)
        }
    }

    private fun handleStatementQuery(text: String): String {
        return interpreter.interpret(text)
    }

    /**
     * Returns error message or null if ok
     */
    private fun handlePosition(text: String) {
        if (handleMultipleLocationsModel(text)) {
            return
        }

        val model = model as SingleLocationBrowserModel
        val locRef = LociCompletion.parse(text, model.genomeQuery)
        if (locRef != null) {
            execute(model.goTo(locRef))
        } else {
            throw IllegalStateException("Illegal location: " + text)
        }
    }

    companion object {
        val GEMLBEE = "GeMLBee - Genome Multiple Locations Browser"

        operator fun invoke(genomeQuery: GenomeQuery, trackViews: ArrayList<TrackView>,
                            interpreter: DesktopInterpreter): DesktopGenomeBrowser {
            return DesktopGenomeBrowser(SingleLocationBrowserModel(genomeQuery),
                    trackViews,
                    LociCompletion.DEFAULT_COMPLETION,
                    interpreter)
        }
    }
}
